File: Backend/src/main/java/com/antigone/rh/service/ClientService.java
package com.antigone.rh.service;

import com.antigone.rh.dto.ClientDTO;
import com.antigone.rh.dto.ClientLoginResponse;
import com.antigone.rh.entity.Client;
import com.antigone.rh.repository.ClientRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.Resource;
import org.springframework.core.io.UrlResource;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.net.MalformedURLException;
import java.nio.file.*;
import java.security.SecureRandom;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
@Slf4j
public class ClientService {

    private final ClientRepository clientRepository;
    private final PasswordEncoder passwordEncoder;
    private final GoogleDriveService googleDriveService;

    @Value("${app.upload.dir:uploads/justificatifs}")
    private String uploadDir;

    private static final String CLIENTS_SUBDIR = "clients";
    private static final SecureRandom RANDOM = new SecureRandom();
    private static final String PASSWORD_CHARS = "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghjkmnpqrstuvwxyz23456789@#$%";

    // ── CRUD ──────────────────────────────────────────────────────────────────

    public List<ClientDTO> getAll() {
        return clientRepository.findAll().stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    public ClientDTO getById(Long id) {
        return toDTO(findOrThrow(id));
    }

    // ── Client portal login ───────────────────────────────────────────────────

    public ClientLoginResponse loginClient(String loginClient, String rawPassword) {
        Client client = clientRepository.findByLoginClient(loginClient)
                .orElseThrow(() -> new RuntimeException("Identifiants invalides"));
        if (client.getPasswordClient() == null ||
                !passwordEncoder.matches(rawPassword, client.getPasswordClient())) {
            throw new RuntimeException("Identifiants invalides");
        }
        // Parse allowed pages
        List<String> pages = parsePages(client.getClientPages());
        return ClientLoginResponse.builder()
                .clientId(client.getId())
                .nom(client.getNom())
                .loginClient(client.getLoginClient())
                .email(client.getEmail())
                .isClient(true)
                .roles(java.util.List.of("CLIENT"))
                .permissions(java.util.List.of("VIEW_CLIENT_PORTAL"))
                .clientPages(pages)
                .build();
    }

    public ClientDTO create(String nom, String email, String telephone, String adresse,
            String notes, String contactNom, String contactPoste, String contactEmail,
            String contactTelephone, boolean createAccount, String clientPages,
            MultipartFile file) throws IOException {

        String filePath = null;
        String fileName = null;
        if (file != null && !file.isEmpty()) {
            fileName = file.getOriginalFilename();
            filePath = saveFile(file);
        }

        // Auto-generate credentials if requested
        String rawPassword = null;
        String loginClient = null;
        String passwordHash = null;
        if (createAccount) {
            loginClient = generateLogin(nom);
            rawPassword = generatePassword(10);
            passwordHash = passwordEncoder.encode(rawPassword);
            log.info("Compte client créé - login={}", loginClient);
        }

        Client client = Client.builder()
                .nom(nom)
                .email(email)
                .telephone(telephone)
                .adresse(adresse)
                .notes(notes)
                .contactNom(contactNom)
                .contactPoste(contactPoste)
                .contactEmail(contactEmail)
                .contactTelephone(contactTelephone)
                .loginClient(loginClient)
                .passwordClient(passwordHash)
                .clientPages(clientPages)
                .filePath(filePath)
                .fileName(fileName)
                .build();

        ClientDTO dto = toDTO(clientRepository.save(client));
        // Return raw password only at creation time
        if (rawPassword != null) {
            dto.setGeneratedPassword(rawPassword);
        }

        // Trigger Async drive hierarchy generation for current year
        googleDriveService.generateFullClientStructure(client.getNom(), java.time.LocalDate.now().getYear());

        return dto;
    }

    public ClientDTO update(Long id, String nom, String email, String telephone, String adresse,
            String notes, String contactNom, String contactPoste, String contactEmail,
            String contactTelephone, boolean regeneratePassword, String clientPages,
            MultipartFile file) throws IOException {

        Client client = findOrThrow(id);

        if (nom != null && !nom.isBlank())
            client.setNom(nom);
        if (email != null)
            client.setEmail(email.isBlank() ? null : email);
        if (telephone != null)
            client.setTelephone(telephone.isBlank() ? null : telephone);
        if (adresse != null)
            client.setAdresse(adresse.isBlank() ? null : adresse);
        if (notes != null)
            client.setNotes(notes.isBlank() ? null : notes);
        if (contactNom != null)
            client.setContactNom(contactNom.isBlank() ? null : contactNom);
        if (contactPoste != null)
            client.setContactPoste(contactPoste.isBlank() ? null : contactPoste);
        if (contactEmail != null)
            client.setContactEmail(contactEmail.isBlank() ? null : contactEmail);
        if (contactTelephone != null)
            client.setContactTelephone(contactTelephone.isBlank() ? null : contactTelephone);

        // Regenerate password if requested (and account already exists)
        String rawPassword = null;
        if (regeneratePassword && client.getLoginClient() != null) {
            rawPassword = generatePassword(10);
            client.setPasswordClient(passwordEncoder.encode(rawPassword));
            log.info("Mot de passe client régénéré - login={}", client.getLoginClient());
        }

        if (file != null && !file.isEmpty()) {
            deleteFile(client.getFilePath());
            client.setFileName(file.getOriginalFilename());
            client.setFilePath(saveFile(file));
        }

        // Always update clientPages (even to null, allowing removal of all access)
        client.setClientPages(clientPages);

        ClientDTO dto = toDTO(clientRepository.save(client));
        if (rawPassword != null) {
            dto.setGeneratedPassword(rawPassword);
        }
        return dto;
    }

    public void delete(Long id) {
        Client client = findOrThrow(id);
        deleteFile(client.getFilePath());
        clientRepository.delete(client);
    }

    // ── File ─────────────────────────────────────────────────────────────────

    public Resource loadFile(Long id) throws MalformedURLException {
        Client client = findOrThrow(id);
        if (client.getFilePath() == null)
            throw new RuntimeException("Aucun fichier attaché.");
        Path path = Paths.get(client.getFilePath()).toAbsolutePath().normalize();
        Resource resource = new UrlResource(path.toUri());
        if (!resource.exists())
            throw new RuntimeException("Fichier introuvable.");
        return resource;
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private String saveFile(MultipartFile file) throws IOException {
        Path dir = Paths.get(uploadDir, CLIENTS_SUBDIR).toAbsolutePath().normalize();
        Files.createDirectories(dir);
        String ext = getExtension(Objects.requireNonNull(file.getOriginalFilename()));
        String uniqueName = UUID.randomUUID() + ext;
        Path target = dir.resolve(uniqueName);
        Files.copy(file.getInputStream(), target, StandardCopyOption.REPLACE_EXISTING);
        return target.toString();
    }

    private void deleteFile(String filePath) {
        if (filePath != null) {
            try {
                Files.deleteIfExists(Paths.get(filePath));
            } catch (IOException ignored) {
            }
        }
    }

    private String getExtension(String filename) {
        int dot = filename.lastIndexOf('.');
        return (dot >= 0) ? filename.substring(dot) : "";
    }

    private Client findOrThrow(Long id) {
        return clientRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Client non trouvé"));
    }

    /**
     * Generate login from client name: lowercase, alphanumeric only, max 15 chars +
     * numeric suffix if collision
     */
    private String generateLogin(String nom) {
        String cleaned = nom.toLowerCase().replaceAll("[^a-z0-9]", "");
        String base = cleaned.isEmpty() ? "client" : cleaned.substring(0, Math.min(cleaned.length(), 15));
        // Ensure uniqueness
        String candidate = base;
        int attempt = 0;
        while (clientRepository.existsByLoginClient(candidate)) {
            candidate = base + (++attempt);
        }
        return candidate;
    }

    private String generatePassword(int length) {
        StringBuilder sb = new StringBuilder(length);
        for (int i = 0; i < length; i++) {
            sb.append(PASSWORD_CHARS.charAt(RANDOM.nextInt(PASSWORD_CHARS.length())));
        }
        return sb.toString();
    }

    public ClientDTO toDTO(Client c) {
        String fileUrl = (c.getFilePath() != null) ? "/api/clients/" + c.getId() + "/file" : null;
        return ClientDTO.builder()
                .id(c.getId())
                .nom(c.getNom())
                .email(c.getEmail())
                .telephone(c.getTelephone())
                .adresse(c.getAdresse())
                .notes(c.getNotes())
                .contactNom(c.getContactNom())
                .contactPoste(c.getContactPoste())
                .contactEmail(c.getContactEmail())
                .contactTelephone(c.getContactTelephone())
                .hasAccount(c.getLoginClient() != null)
                .loginClient(c.getLoginClient())
                .clientPages(c.getClientPages())
                // passwordClient is hashed — never expose it in DTO except generatedPassword
                .description(c.getDescription())
                .responsable(c.getResponsable())
                .fileName(c.getFileName())
                .fileUrl(fileUrl)
                .dateCreation(c.getDateCreation())
                .build();
    }

    /** Parses "MEDIA_PLANS,PROJETS,FICHIERS" format into a List. */
    private List<String> parsePages(String clientPages) {
        if (clientPages == null || clientPages.isBlank()) return Collections.emptyList();
        return Arrays.stream(clientPages.split(","))
                .map(String::trim)
                .filter(s -> !s.isEmpty())
                .collect(Collectors.toList());
    }
}

File: frontend-projects/src/api/clientService.ts
import api from './axios';

export interface ClientDTO {
    id: number;
    nom: string;
    email?: string;
    telephone?: string;
    adresse?: string;
    notes?: string;
    // Contact principal
    contactNom?: string;
    contactPoste?: string;
    contactEmail?: string;
    contactTelephone?: string;
    // Account
    hasAccount?: boolean;
    loginClient?: string;
    /** Comma-separated page keys (e.g. "MEDIA_PLANS,PROJETS,FICHIERS") */
    clientPages?: string;
    /** Only returned once at creation */
    generatedPassword?: string;
    // Legacy
    description?: string;
    responsable?: string;
    fileName?: string;
    fileUrl?: string;
    dateCreation: string;
}

const BASE = '/clients';

const getAllClients = () => api.get<{ data: ClientDTO[] }>(BASE);
const getClientById = (id: number) => api.get<{ data: ClientDTO }>(`${BASE}/${id}`);
const getClientPortalDriveLink = (id: number) => api.get<{ data: string }>(`${BASE}/${id}/client-portal-drive-link`);
const getDriveFiles = (id: number) => api.get<{ data: MonthFilesGroup[] }>(`${BASE}/${id}/drive-files`);

export interface DriveFile {
    id: string;
    name: string;
    mimeType: string;
    webViewLink: string;
    thumbnailLink?: string;
    size?: number;
    modifiedTime?: string;
    subFolder?: string;
}

export interface MonthFilesGroup {
    monthLabel: string;
    files: DriveFile[];
}

const createClient = (data: {
    nom: string;
    email?: string;
    telephone?: string;
    adresse?: string;
    notes?: string;
    contactNom?: string;
    contactPoste?: string;
    contactEmail?: string;
    contactTelephone?: string;
    createAccount?: boolean;
    clientPages?: string;
    file?: File;
}) => {
    const form = new FormData();
    form.append('nom', data.nom);
    if (data.email) form.append('email', data.email);
    if (data.telephone) form.append('telephone', data.telephone);
    if (data.adresse) form.append('adresse', data.adresse);
    if (data.notes) form.append('notes', data.notes);
    if (data.contactNom) form.append('contactNom', data.contactNom);
    if (data.contactPoste) form.append('contactPoste', data.contactPoste);
    if (data.contactEmail) form.append('contactEmail', data.contactEmail);
    if (data.contactTelephone) form.append('contactTelephone', data.contactTelephone);
    form.append('createAccount', String(data.createAccount ?? false));
    if (data.clientPages != null) form.append('clientPages', data.clientPages);
    if (data.file) form.append('file', data.file);
    return api.post<{ data: ClientDTO }>(BASE, form, {
        headers: { 'Content-Type': 'multipart/form-data' },
    });
};

const updateClient = (id: number, data: {
    nom?: string;
    email?: string;
    telephone?: string;
    adresse?: string;
    notes?: string;
    contactNom?: string;
    contactPoste?: string;
    contactEmail?: string;
    contactTelephone?: string;
    regeneratePassword?: boolean;
    clientPages?: string;
    file?: File;
}) => {
    const form = new FormData();
    if (data.nom) form.append('nom', data.nom);
    form.append('email', data.email ?? '');
    form.append('telephone', data.telephone ?? '');
    form.append('adresse', data.adresse ?? '');
    form.append('notes', data.notes ?? '');
    form.append('contactNom', data.contactNom ?? '');
    form.append('contactPoste', data.contactPoste ?? '');
    form.append('contactEmail', data.contactEmail ?? '');
    form.append('contactTelephone', data.contactTelephone ?? '');
    form.append('regeneratePassword', String(data.regeneratePassword ?? false));
    if (data.clientPages != null) form.append('clientPages', data.clientPages);
    if (data.file) form.append('file', data.file);
    return api.put<{ data: ClientDTO }>(`${BASE}/${id}`, form, {
        headers: { 'Content-Type': 'multipart/form-data' },
    });
};

const deleteClient = (id: number) => api.delete(`${BASE}/${id}`);

export const clientService = {
    getAllClients,
    getClientById,
    getClientPortalDriveLink,
    getDriveFiles,
    createClient,
    updateClient,
    deleteClient,
};
